File: infrastructure/go-sdk/dapr.go
package infra

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"go.uber.org/zap"
)

type DaprClient struct {
	httpPort   int
	grpcPort   int
	httpClient *http.Client
	logger     *zap.Logger
}

const (
	DaprStateStore   = "statestore"
	DaprPubSub       = "pubsub"
	DaprSecretStore  = "secretstore"
)

func NewDaprClient(logger *zap.Logger, httpPort int) *DaprClient {
	if httpPort == 0 {
		httpPort = 3500
	}
	return &DaprClient{
		httpPort:   httpPort,
		grpcPort:   50001,
		httpClient: &http.Client{Timeout: 5 * time.Second},
		logger:     logger,
	}
}

func (c *DaprClient) baseURL() string {
	return fmt.Sprintf("http://localhost:%d", c.httpPort)
}

func (c *DaprClient) Ping(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, "GET", c.baseURL()+"/v1.0/healthz", nil)
	if err != nil {
		return err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("dapr ping: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != 204 && resp.StatusCode != 200 {
		return fmt.Errorf("dapr unhealthy: %d", resp.StatusCode)
	}
	return nil
}

// SaveState saves key-value state with optional ETag for optimistic concurrency.
func (c *DaprClient) SaveState(ctx context.Context, storeName, key string, value interface{}, etag string) error {
	item := map[string]interface{}{
		"key":   key,
		"value": value,
	}
	if etag != "" {
		item["etag"] = etag
		item["options"] = map[string]string{"concurrency": "first-write"}
	}
	data, _ := json.Marshal([]interface{}{item})
	req, err := http.NewRequestWithContext(ctx, "POST", fmt.Sprintf("%s/v1.0/state/%s", c.baseURL(), storeName), bytes.NewReader(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("save state: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("save state failed (%d): %s", resp.StatusCode, string(body))
	}
	return nil
}

// GetState retrieves state by key, returning the value and ETag.
func (c *DaprClient) GetState(ctx context.Context, storeName, key string) ([]byte, string, error) {
	req, err := http.NewRequestWithContext(ctx, "GET", fmt.Sprintf("%s/v1.0/state/%s/%s", c.baseURL(), storeName, key), nil)
	if err != nil {
		return nil, "", err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, "", fmt.Errorf("get state: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode == 204 || resp.StatusCode == 404 {
		return nil, "", nil
	}
	body, _ := io.ReadAll(resp.Body)
	etag := resp.Header.Get("ETag")
	return body, etag, nil
}

// DeleteState deletes a key from the state store.
func (c *DaprClient) DeleteState(ctx context.Context, storeName, key string) error {
	req, err := http.NewRequestWithContext(ctx, "DELETE", fmt.Sprintf("%s/v1.0/state/%s/%s", c.baseURL(), storeName, key), nil)
	if err != nil {
		return err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("delete state: %w", err)
	}
	defer resp.Body.Close()
	return nil
}

// PublishEvent publishes an event to a pub/sub topic.
func (c *DaprClient) PublishEvent(ctx context.Context, pubsubName, topic string, event interface{}) error {
	data, _ := json.Marshal(event)
	req, err := http.NewRequestWithContext(ctx, "POST", fmt.Sprintf("%s/v1.0/publish/%s/%s", c.baseURL(), pubsubName, topic), bytes.NewReader(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("publish event: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("publish failed (%d): %s", resp.StatusCode, string(body))
	}
	return nil
}

// InvokeService invokes another service via Dapr service invocation.
func (c *DaprClient) InvokeService(ctx context.Context, appID, method string, payload interface{}) ([]byte, error) {
	var bodyReader io.Reader
	if payload != nil {
		data, _ := json.Marshal(payload)
		bodyReader = bytes.NewReader(data)
	}
	req, err := http.NewRequestWithContext(ctx, "POST", fmt.Sprintf("%s/v1.0/invoke/%s/method/%s", c.baseURL(), appID, method), bodyReader)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("invoke %s/%s: %w", appID, method, err)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("invoke failed (%d): %s", resp.StatusCode, string(body))
	}
	return body, nil
}

// GetSecret retrieves a secret from a secret store.
func (c *DaprClient) GetSecret(ctx context.Context, storeName, key string) (map[string]string, error) {
	req, err := http.NewRequestWithContext(ctx, "GET", fmt.Sprintf("%s/v1.0/secrets/%s/%s", c.baseURL(), storeName, key), nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("get secret: %w", err)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	var secrets map[string]string
	json.Unmarshal(body, &secrets)
	return secrets, nil
}

// SaveKYCSession stores a KYC session in Dapr state.
func (c *DaprClient) SaveKYCSession(ctx context.Context, sessionID string, session interface{}) error {
	return c.SaveState(ctx, DaprStateStore, "kyc:session:"+sessionID, session, "")
}

// GetKYCSession retrieves a KYC session from Dapr state.
func (c *DaprClient) GetKYCSession(ctx context.Context, sessionID string) ([]byte, error) {
	data, _, err := c.GetState(ctx, DaprStateStore, "kyc:session:"+sessionID)
	return data, err
}

// PublishKYCEvent publishes a KYC event via Dapr pub/sub.
func (c *DaprClient) PublishKYCEvent(ctx context.Context, eventType string, payload interface{}) error {
	event := map[string]interface{}{
		"event_type": eventType,
		"data":       payload,
		"timestamp":  time.Now().UTC().Format(time.RFC3339),
		"source":     "kyc-orchestrator",
	}
	return c.PublishEvent(ctx, DaprPubSub, "kyc-events", event)
}

// InvokeKYCService invokes another KYC-related service via Dapr.
func (c *DaprClient) InvokeKYCService(ctx context.Context, service, method string, payload interface{}) ([]byte, error) {
	return c.InvokeService(ctx, service, method, payload)
}

// SavePolicyState stores policy state in Dapr.
func (c *DaprClient) SavePolicyState(ctx context.Context, policyID string, state interface{}) error {
	return c.SaveState(ctx, DaprStateStore, "policy:"+policyID, state, "")
}

// SaveClaimState stores claim state in Dapr.
func (c *DaprClient) SaveClaimState(ctx context.Context, claimID string, state interface{}) error {
	return c.SaveState(ctx, DaprStateStore, "claim:"+claimID, state, "")
}

File: kyc-kyb-system/kyc-ledger-service/src/dapr.rs
use serde::{Deserialize, Serialize, de::DeserializeOwned};
use reqwest::Client;

pub struct DaprClient {
    base_url: String,
    client: Client,
}

impl DaprClient {
    pub fn new(dapr_port: &str) -> Self {
        Self {
            base_url: format!("http://localhost:{}", dapr_port),
            client: Client::new(),
        }
    }

    pub fn is_connected(&self) -> bool {
        true
    }

    pub async fn save_state<T: Serialize>(&self, store_name: &str, key: &str, value: &T) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("{}/v1.0/state/{}", self.base_url, store_name);
        let payload = serde_json::json!([{
            "key": key,
            "value": value,
        }]);

        let resp = self.client.post(&url).json(&payload).send().await;
        match resp {
            Ok(r) if r.status().is_success() => {
                tracing::info!(store = store_name, key = key, "dapr_state_saved");
                Ok(())
            }
            Ok(r) => {
                tracing::warn!(status = %r.status(), "dapr_state_save_warning");
                Ok(())
            }
            Err(e) => {
                tracing::debug!(error = %e, "dapr_not_available");
                Ok(())
            }
        }
    }

    pub async fn get_state<T: DeserializeOwned>(&self, store_name: &str, key: &str) -> Result<Option<T>, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("{}/v1.0/state/{}/{}", self.base_url, store_name, key);
        let resp = self.client.get(&url).send().await;

        match resp {
            Ok(r) if r.status().is_success() => {
                match r.json::<T>().await {
                    Ok(val) => Ok(Some(val)),
                    Err(_) => Ok(None),
                }
            }
            _ => Ok(None),
        }
    }

    pub async fn delete_state(&self, store_name: &str, key: &str) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("{}/v1.0/state/{}/{}", self.base_url, store_name, key);
        let _ = self.client.delete(&url).send().await;
        Ok(())
    }

    pub async fn publish_event<T: Serialize>(&self, pubsub_name: &str, topic: &str, data: &T) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("{}/v1.0/publish/{}/{}", self.base_url, pubsub_name, topic);
        let resp = self.client.post(&url)
            .header("Content-Type", "application/json")
            .json(data)
            .send()
            .await;

        match resp {
            Ok(r) if r.status().is_success() => {
                tracing::info!(pubsub = pubsub_name, topic = topic, "dapr_event_published");
                Ok(())
            }
            Ok(r) => {
                tracing::warn!(status = %r.status(), "dapr_publish_warning");
                Ok(())
            }
            Err(e) => {
                tracing::debug!(error = %e, "dapr_pubsub_not_available");
                Ok(())
            }
        }
    }

    pub async fn invoke_service(&self, app_id: &str, method: &str, data: &serde_json::Value) -> Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("{}/v1.0/invoke/{}/method/{}", self.base_url, app_id, method);
        let resp = self.client.post(&url).json(data).send().await?;

        if resp.status().is_success() {
            Ok(resp.json().await?)
        } else {
            Err(format!("Dapr invoke failed: {}", resp.status()).into())
        }
    }

    pub async fn get_secret(&self, store_name: &str, key: &str) -> Result<Option<String>, Box<dyn std::error::Error + Send + Sync>> {
        let url = format!("{}/v1.0/secrets/{}/{}", self.base_url, store_name, key);
        let resp = self.client.get(&url).send().await;

        match resp {
            Ok(r) if r.status().is_success() => {
                let secrets: std::collections::HashMap<String, String> = r.json().await?;
                Ok(secrets.get(key).cloned())
            }
            _ => Ok(None),
        }
    }

    pub async fn check_kyc_gate(&self, user_id: &str) -> Result<(bool, u8), Box<dyn std::error::Error + Send + Sync>> {
        let result = self.invoke_service(
            "kyc-orchestrator",
            &format!("api/v1/kyc/gate/{}", user_id),
            &serde_json::json!({}),
        ).await;

        match result {
            Ok(data) => {
                let allowed = data.get("allowed").and_then(|v| v.as_bool()).unwrap_or(false);
                let level = data.get("level").and_then(|v| v.as_u64()).unwrap_or(0) as u8;
                Ok((allowed, level))
            }
            Err(_) => Ok((false, 0)),
        }
    }
}
